simulators/portal/src/suites/beacon/sync.rs

```diff
@@ -0,0 +1,203 @@
+use hivesim::{dyn_async, Client, NClientTestSpec, Test};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::time::{sleep, Duration};
+
+dyn_async! {
+   pub async fn test_beacon_sync<'a> (test: &'a mut Test, _client: Option<Client>) {
+        // Get all available portal clients
+        let clients = test.sim.client_types().await;
+
+        // this is the "blank" client, used just for storing beacon network
+        // syncing data, and then we test sync functionality on the other client
+        let environment = Some(HashMap::from([
+            PortalNetwork::as_environment_flag([PortalNetwork::Beacon]),
+        ]));
+        let environments = Some(vec![environment.clone(), environment]);
+
+        // Iterate over all possible pairings of clients and run the tests (including self-pairings)
+        for client in &clients {
+            test.run(
+                NClientTestSpec {
+                    name: format!("Beacon sync test: latest finalized root{}", client.name),
+                    description: "".to_string(),
+                    always_run: false,
+                    run: test_client_syncs_with_latest_finalized_root,
+                    environments: environments.clone(),
+                    test_data: (),
+                    clients: vec![client.clone()],
+                }
+            ).await;
+            test.run(
+                NClientTestSpec {
+                    name: format!("Beacon sync test: latest optimistic root. {}", client.name),
+                    description: "".to_string(),
+                    always_run: false,
+                    run: test_client_syncs_with_latest_optimistic_root,
+                    environments: environments.clone(),
+                    test_data: (),
+                    clients: vec![client.clone()],
+                }
+            ).await;
+        }
+   }
+}
+
+dyn_async! {
+    async fn test_client_syncs_with_latest_finalized_root<'a>(clients: Vec<Client>, _: ()) {
+        let Some((client)) = clients.into_iter().next() else {
+            panic!("Unable to get expected amount of clients from NClientTestSpec")
+        };
+
+        // starts the bridge service:
+        // the bridge service acts like a beacon bridge, but simply injects
+        // the latest data into the given client, so that it will be available
+        // for syncing from the client we're testing
+        let bridge_service = Arc::new(BridgeService::new(client.clone()));
+        let service_for_task = bridge_service.clone();
+        let provider_handle = tokio::spawn(async move {
+            service_for_task.start().await;
+        });
+
+        // get enr
+        let client_enr = match client.rpc.node_info().await {
+            Ok(node_info) => node_info.enr,
+            Err(err) => panic!("Error getting node info: {err:?}"),
+        };
+
+        // wait for initial trusted block root
+        let mut trusted_block_root = None;
+        for i in 0..30 {
```

Comment:
We can use tokio timeout() for an overall deadline rather than counting retries and instead of counting 30 attempts, we can say “give me a root within 30 seconds or fail.”:

```rust
use tokio::time::{timeout, Duration, sleep};

let result = timeout(Duration::from_secs(30), async {
    loop {
        if let Some(trusted_root) = bridge_service.trusted_block_root().await {
            return trusted_root; 
        }
        sleep(Duration::from_secs(1)).await;
    }
}).await;

// If `timeout` expired, handle the error:
match result {
    Ok(trusted_root) => {
        // We got the root before 30s
        // do something with `trusted_root`
    }
    Err(_) => {
        drop(provider_handle);
        panic!("Trusted block root not initialized in time");
    }
}
```